src/bgw/job.c

```diff
@@ -0,0 +1,317 @@
+
+	MemoryContextSwitchTo(orig);
+	return false;
+}
+
+
+extern List *
+bgw_job_get_all(size_t alloc_size, MemoryContext mctx)
+{
+	Catalog    *catalog = catalog_get();
+	AccumData	list_data = {
+		.list = NIL,
+		.alloc_size = alloc_size,
+	};
+	int			num_tuples;
+	ScannerCtx	scanctx = {
+		.table = catalog->tables[BGW_JOB].id,
+		.index = InvalidOid,
+		.data = &list_data,
+		.tuple_found = bgw_job_accum_tuple_found,
+		.lockmode = AccessShareLock,
+		.scandirection = ForwardScanDirection,
+		.result_mctx = mctx,
+	};
+
+	num_tuples = scanner_scan(&scanctx);
+	Assert(list_length(list_data.list) == num_tuples);
+	return list_data.list;
+}
+
+static void
+bgw_job_scan_one(int indexid, ScanKeyData scankey[], int nkeys,
+				 tuple_found_func tuple_found, tuple_filter_func tuple_filter, void *data, MemoryContext mctx, LOCKMODE lockmode)
+{
+	Catalog    *catalog = catalog_get();
+	ScannerCtx	scanCtx = {
+		.table = catalog->tables[BGW_JOB].id,
+		.index = CATALOG_INDEX(catalog, BGW_JOB, indexid),
+		.nkeys = nkeys,
+		.scankey = scankey,
+		.tuple_found = tuple_found,
+		.filter = tuple_filter,
+		.data = data,
+		.lockmode = lockmode,
+		.scandirection = ForwardScanDirection,
+		.result_mctx = mctx,
+	};
+
+	scanner_scan_one(&scanCtx, true, "bgw job");
+}
+
+static inline void
+bgw_job_scan_job_id(int32 bgw_job_id, tuple_found_func tuple_found, tuple_filter_func tuple_filter,
+					void *data, MemoryContext mctx, LOCKMODE lockmode)
+{
+	ScanKeyData scankey[1];
+
+	ScanKeyInit(&scankey[0],
+				Anum_bgw_job_stat_pkey_idx_job_id,
+				BTEqualStrategyNumber, F_INT4EQ, Int32GetDatum(bgw_job_id));
+	bgw_job_scan_one(BGW_JOB_STAT_PKEY_IDX,
+					 scankey, 1, tuple_found, tuple_filter, data, mctx, lockmode);
+}
+
+static bool
+bgw_job_tuple_found(TupleInfo *ti, void *const data)
+{
+	BgwJob	  **job_pp = data;
+
+	*job_pp = bgw_job_from_tuple(ti->tuple, sizeof(BgwJob), ti->mctx);
+
+	/* return true since used with scan one */
+	return true;
+}
+
+BgwJob *
+bgw_job_find(int32 bgw_job_id, MemoryContext mctx)
+{
+	BgwJob	   *job_stat = NULL;
+
+	bgw_job_scan_job_id(bgw_job_id, bgw_job_tuple_found, NULL, &job_stat, mctx, AccessShareLock);
+
+	return job_stat;
+}
+
+static
+JobType
+get_job_type_from_name(Name job_type_name)
+{
+	for (int i = 0; i < _MAX_JOB_TYPE; i++)
+	{
+		if (namestrcmp(job_type_name, job_type_names[i]) == 0)
+			return i;
+	}
+	return _MAX_JOB_TYPE;
+}
+
+#ifdef DEBUG
+#define TEST_JOB_FUNCTION "ts_bgw_job_execute_test"
+static bool
+call_test_dispatcher(BgwJob *job)
+{
+	/* the test dispatcher should throw an error if function not found */
+	PGFunction test_function = load_external_function(extension_get_so_name(), TEST_JOB_FUNCTION, false, NULL);
+
+	if (test_function == NULL)
+		elog(ERROR, "cannot find test execution function");
+
+	return DatumGetBool(DirectFunctionCall1(test_function, Int32GetDatum(job->fd.id)));
+}
+#endif
+
+bool
+bgw_job_execute(BgwJob *job)
+{
+	switch (get_job_type_from_name(&job->fd.job_type))
```

Comment:
A switch is fine, but if the function interface is the same for all jobs, you can just create a static type-to-function array and index that as an alternative approach.

Also, see my note at the definition of `BgwJob` on caching the type in the struct so that you can do `job->type`.

Reply:
Added caching. I like the switch more for now.